pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Map;
+import org.apache.pinot.common.concurrency.AdjustableSemaphore;
+import org.apache.pinot.spi.config.provider.PinotClusterConfigChangeListener;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Used to throttle the total concurrent index rebuilds that can happen on a given Pinot server.
+ * Code paths that do no need to rebuild the index or which don't happen on the server need not utilize this throttler.
+ */
+public class SegmentPreprocessThrottler implements PinotClusterConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessThrottler.class);
+
+  /**
+   * If _maxPreprocessConcurrency is <= 0, it means that the cluster is not configured to limit the number of
+   * index rebuilds that can be executed concurrently. This class sets the semaphore permit to be a large value
+   * in this scenario.
+   */
+  private int _maxPreprocessConcurrency;
+  private boolean _relaxThrottling;
+  private volatile AdjustableSemaphore _semaphore;
+
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, boolean relaxThrottling) {
+    LOGGER.info("Initializing SegmentPreprocessThrottler, maxPreprocessConcurrency: {}, relaxThrottling: {}",
+        maxPreprocessConcurrency, relaxThrottling);
+    if (maxPreprocessConcurrency <= 0) {
+      int defaultSegmentPreprocessParallelism =
+          Integer.parseInt(CommonConstants.Helix.DEFAULT_MAX_SEGMENT_PREPROCESS_PARALLELISM);
+      LOGGER.warn("Max preprocess concurrency is negative: {}, resetting to {}", maxPreprocessConcurrency,
+          defaultSegmentPreprocessParallelism);
+      maxPreprocessConcurrency = defaultSegmentPreprocessParallelism;
+    }
+    _maxPreprocessConcurrency = maxPreprocessConcurrency;
+    _relaxThrottling = relaxThrottling;
+
+    int relaxThrottlingThreshold = Math.max(_maxPreprocessConcurrency,
+        CommonConstants.Helix.DEFAULT_MAX_SEGMENT_PREPROCESS_PARALLELISM_BEFORE_SERVING_QUERIES);
+    int preprocessConcurrency = _maxPreprocessConcurrency;
+    if (relaxThrottling) {
+      preprocessConcurrency = relaxThrottlingThreshold;
+      LOGGER.info("Relax throttling enabled, setting preprocess concurrency to: {}", preprocessConcurrency);
+    }
+    _semaphore = new AdjustableSemaphore(preprocessConcurrency, true);
+    LOGGER.info("Created semaphore with available permits: {}", _semaphore.availablePermits());
+  }
+
+  public void resetThrottling() {
```

Comment:
Will there be race condition between this method and `onChange()`?

Reply:
good catch, i've made both of them synchronized. this will only be called once, but even then it can coincide with onChange and miss the updated value.